### ISSUE TITLE
[BUG] [Spark] Documentation is ambiguous - Multi-cluster writes in GCS 

### ISSUE THREAD

**Issue by joydeepbroy-zeotap** [delta-io/delta#2455]
#### Which Delta project/connector is this regarding?

- [X] Spark
- [ ] Standalone
- [ ] Flink
- [ ] Kernel
- [ ] Other (fill in here)

### Describe the problem
**Multi-Cluster Write Issue on GCS with delta-lake v1.0.1**
We had two delta lake jobs from different spark clusters (basically a multi-cluster setup), both of them updating the same delta table but operating on completely different partitions. We found that both of them committed the same delta version which resulted in overwriting the Delta Version Log by one job which was already written by the other. This is a similar issue alluded to by @dennyglee in the blog post for AWS S3: https://delta.io/blog/2022-05-18-multi-cluster-writes-to-delta-lake-storage-in-s3/

#### Observed results

We checked various documentation from both GCP and Delta Lake:
- https://cloud.google.com/storage/docs/consistency
- https://docs.delta.io/latest/delta-storage.html#google-cloud-storage
- https://cloud.google.com/blog/products/data-analytics/getting-started-with-new-table-formats-on-dataproc/

Many of them claim that the 3 guarantees of atomic visibility, mutual exclusion, and consistent listing are already supported by GCP, and hence, no separate `LogStore` implementation is required. However, if that was the case then the issue should not have happened with v1.0. Finally, we stumbled upon a conversation on issue #294 which clarifies that GCS does not support the issue, and PR #560 was raised in v2.0.0 to solve the issue.

We then turned to the documentation for help with the matter. The [GCS](https://docs.delta.io/latest/delta-storage.html#google-cloud-storage) section of the documentation seem to be ambiguous.

- Our conclusion from the blogs and issues is that Multi-Cluster writes for GCS work only beyond v2.0 and do not work for v1.0. Is this analysis correct?
- It states that

> For Delta Lake 1.2.0 and below, you must explicitly configure the LogStore implementation for the scheme gs. as spark.delta.logStore.gs.impl=io.delta.storage.GCSLogStore. 

 - - However, the GCSLogStore class was only introduced in v2.0.0, then how would it resolve the class in a version lower than that? In fact when we launched a job with delta v1.0.1 and passed the above config, it threw a `java.lang.ClassNotFoundException: io.delta.storage.GCSLogStore`

- It further states `Include the JAR for gcs-connector in the classpath`. We checked in our classpath for the `gcs-connector` jar and indeed it was present. But, that did not prevent the issue.

#### Expected results

1. First of all, we would like to know if our analysis is correct or are there any gaps in what we understood?
2. For AWS S3 there is a clear warning for Muli-Cluster writes and a suggestion to use DynamoDB. If our analysis is correct, then a similar warning should be clearly stated in the documentation that Multi-Cluster write for GCS only works beyond v2.0.0.
3. Furthermore, even if this analysis is incorrect, there should be some clear documentation around multi-cluster writes on GCS in general as there are a lot of ambiguous blogs that gives the impression that GCS supports multi-cluster writes with Delta Lake regardless of the version. From Delta Lake's end, there is just the one issue #294  which talks about it. Since this can cause data loss in production, at least there should be a warning pertaining to it.
4. If the `GCSLogStore` argument is correct, then that configuration should either be removed or fixed to show the correct version

### Environment information

* Delta Lake version: `io.delta:delta-core_2.12:1.0.1`
* Spark version: `3.1.3`
* Scala version: `2.12`

### Willingness to contribute

The Delta Lake Community encourages bug fix contributions. Would you or another member of your organization be willing to contribute a fix for this bug to the Delta Lake code base?

- [x] Yes. I can contribute a fix for this bug independently.
- [ ] Yes. I would be willing to contribute a fix for this bug with guidance from the Delta Lake community.
- [ ] No. I cannot contribute a bug fix at this time.

**Comment by scottsand-db** [comment]
Hi @joydeepbroy-zeotap. Thanks for your questions. I'll do my best to answer.

## 1
First, why are you using Delta Lake 1.0.1. Can you use a newer version? Are you limited by your version of Spark?

## 2
> We had two delta lake jobs from different spark clusters (basically a multi-cluster setup), both of them updating the same delta table but operating on completely different partitions. We found that both of them committed the same delta version which resulted in overwriting the Delta Version Log by one job which was already written by the other.

FYI, physical concurrency issues only occur when committing to the delta log (e.g. two writers both want to write `008.json` into the delta log). For phsyical concurrency, it doesn't matter which partitions they wrote to btw.

## 3
> For Delta Lake 1.2.0 and below, you must explicitly configure the LogStore implementation for the scheme gs. as spark.delta.logStore.gs.impl=io.delta.storage.GCSLogStore.
> However, the GCSLogStore class was only introduced in v2.0.0, then how would it resolve the class in a version lower than that? In fact when we launched a job with delta v1.0.1 and passed the above config, it threw a java.lang.ClassNotFoundException: io.delta.storage.GCSLogStore

Did you follow: `For Delta Lake 1.1 and below, you must explicitly include the JAR of the Delta Contributions (delta-contribs) Maven artifact of the same version.`. The class is there in version 1.0.1: https://github.com/delta-io/delta/blob/v1.0.1/contribs/src/main/scala/io/delta/storage/GCSLogStore.scala

## 4
> Many of them claim that the 3 guarantees of atomic visibility, mutual exclusion, and consistent listing are already supported by GCP, and hence, no separate LogStore implementation is required. 

Err, no. As our docs say, `You must configure Delta Lake to use the correct LogStore for concurrently reading and writing from GCS`

You must use the GCSLogStore. Please configure it with: `spark.delta.logStore.gs.impl=io.delta.storage.GCSLogStore`

## 5
> We found that both of them committed the same delta version which resulted in overwriting the Delta Version Log by one job which was already written by the other. 

My best guess is that this was caused by the GCSLogStore not being properly configured for all concurrent writers to that table.

**Comment by joydeepbroy-zeotap** [comment]
Thanks @scottsand-db for the detailed explanation. 

You are right that the delta-contribs jar was missing and wasn't included in the spark job. We followed the link provided on the [GCS documentation](https://docs.delta.io/latest/delta-storage.html#google-cloud-storage) page and it redirected to this page : https://central.sonatype.com/search?q=g:io.delta%20%20a:delta-contribs*&smo=true

It was showing zero results but we should have investigated further and figured out where the delta-contribs jar resided, which we actually found here: https://central.sonatype.com/search?q=delta-contribs&smo=true

Maybe you guys can fix the Maven link on the page. But yes, it works fine now and we can load the `GCSLogStore` with delta 1.0.1 in spark-shell or spark-submit. Additionally, we moved to delta 2.x after the incident where GCSLogStore is part of the main project. So, we can't test the same workload for Multi-Cluster-Write but grateful for the help nevertheless.
